### PR TITLE
Refactor "sequence" to "routes"

### PR DESCRIPTION
**Pull request by sanderPostma**
- useFlowRouter() instead of `useState<Sequencer>`
- mapped config to stepId instead of static 1:1 config to page relation. This allows for using the same page FC's with different configurations for different routes/steps.
- separated PageOrComponentConfig into PageConfig & ComponentConfig

### CODE REVIEW

**Comment by sanderPostma** [comment]
Branch active on https://kvk.demo.sphereon.com for testing

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialsLandingPage/index.tsx [hunk 1]
```diff
@@ -2,12 +2,14 @@ import React from 'react';
...
 import {
-    getCurrentEcosystemPageOrComponentConfig,
+    
     SSICredentialCardConfig,
     SSICredentialsLandingPageConfig
 } from "../../ecosystem-config";
 import {Mobile, NonMobile} from "../../index";
 import {useMediaQuery} from "react-responsive";
+import {useFlowRouter} from "../../router/flow-router"
+import {useLocation} from "react-router-dom"
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSICredentialsLandingPage/index.tsx]: not used import

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialIssuedSuccessPage/index.tsx [hunk 2]
```diff
@@ -4,28 +4,21 @@ import {useTranslation} from "react-i18next";
...
 import {
     getCurrentEcosystemGeneralConfig,
-    getCurrentEcosystemPageOrComponentConfig,
+    
     SSICredentialIssuedSuccessPageConfig
 } from "../../ecosystem-config";
 import {NonMobile} from "../../index";
 import {useMediaQuery} from "react-responsive";
-import SSISecondaryButton from "../../components/SSISecondaryButton";
 import {useLocation, useNavigate} from "react-router-dom"
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSICredentialIssuedSuccessPage/index.tsx]: not used anymore, also the same for `useEffect, useState`

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx [hunk 3]
```diff
@@ -1,40 +1,36 @@
...
     getCurrentEcosystemGeneralConfig,
-    getCurrentEcosystemPageOrComponentConfig,
     SSICredentialVerifyRequestPageConfig
 } from "../../ecosystem-config";
 import SSIPrimaryButton from "../../components/SSIPrimaryButton";
 import {useMediaQuery} from "react-responsive";
 import {Mobile, MobileOS, NonMobile} from "../../index"
-import {Sequencer} from "../../router/sequencer"
+import {useFlowRouter} from "../../router/flow-router"
+import {useLocation} from "react-router-dom"
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: not used

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx [hunk 4]
```diff
@@ -1,40 +1,36 @@
-import React, {ReactElement, useState} from 'react'
+import React, {ReactElement, useEffect, useState} from 'react'
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: reactelement and useeffect not used

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx [hunk 5]
```diff
@@ -119,7 +115,7 @@ export default function SSICredentialVerifyRequestPage(): React.ReactElement | n
                             <SSIPrimaryButton
                                 caption={t('credential_verify_request_right_pane_button_caption')}
                                 onClick={async () => {
-                                    sequencer.goToStep(config.rightPaneButtonStepId ?? 'infoRequest')
+                                    flowRouter.goToStep(config.rightPaneButtonStepId ?? 'infoRequest')
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: await

File: packages/oid4vci-demo-frontend/src/components/SSISecondaryButton/index.tsx [hunk 6]
```diff
@@ -1,9 +1,9 @@
 import React from 'react';
 import {
-    getCurrentEcosystemPageOrComponentConfig,
+    getCurrentEcosystemComponentConfig,
     SSIPrimaryButtonConfig,
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/components/SSISecondaryButton/index.tsx:4]: `SSIPrimaryButtonConfig` not used

File: packages/oid4vci-demo-frontend/src/pages/SSIInformationRequestPage/index.tsx [hunk 7]
```diff
@@ -15,17 +15,17 @@ import {
...
-  getCurrentEcosystemPageOrComponentConfig,
+  
   SSIInformationRequestPageConfig
 } from "../../ecosystem-config"
 import SSIPrimaryButton from "../../components/SSIPrimaryButton";
-import {useLocation, useNavigate} from "react-router-dom"
+import {useLocation} from "react-router-dom"
 import {Buffer} from 'buffer';
 import {useMediaQuery} from "react-responsive";
 import {Mobile, NonMobile} from "../../index";
```
Comment (sksadjad) [inline packages/oid4vci-demo-frontend/src/pages/SSIInformationRequestPage/index.tsx]: mobile not used

**Review by sksadjad — CHANGES_REQUESTED** [review]
added a few notes

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx (hunk 5 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: @sksadjad is there a point to add await if you don't have any follow-up code that has to wait for it?

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialsLandingPage/index.tsx (hunk 1 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/pages/SSICredentialsLandingPage/index.tsx]: I let IntelliJ do "Optimize imports" on the pages and the router, so all imports should be clean now

File: packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx (hunk 5 above)
Comment (sksadjad) [reply packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: in general I think it's better to wait for your async function. it can cause resource issues, it can wait for it indefinitely or other problems/issues

Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/pages/SSICredentialVerifyRequestPage/index.tsx]: fixed

**Review by sksadjad — APPROVED** [review]
looks good to me

File: packages/oid4vci-demo-frontend/src/ecosystem-config.ts [hunk 8]
```diff
@@ -14,41 +14,53 @@ export function getCurrentEcosystemConfig(): VCIConfig {
...
     }
-    if (pageOrComponent in config.pages) {
-        return config.pages[pageOrComponent as keyof VCIConfigPages]
-    } else if (pageOrComponent in config.components) {
-        return config.components[pageOrComponent as keyof VCIConfigComponents]
+    if (stepId in config.pages) {
+        return config.pages[stepId as keyof VCIConfigPages]
     }
-    throw new Error(`config for ${pageOrComponent} doesn't exist`)
+    return undefined
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/ecosystem-config.ts]: Hmzz, why is this changed from error to undefined? Other similar functions also return errors.

File: packages/oid4vci-demo-frontend/src/pages/SSIDownloadPage/index.tsx [hunk 9]
```diff
@@ -1,31 +1,20 @@
...
+import {useMediaQuery} from "react-responsive"
+import {useFlowRouter} from "../../router/flow-router"
 
 const SSIDownloadPage: React.FC = () => {
-    const [sequencer] = useState<Sequencer>(new Sequencer())
-    const location = useLocation()
-    const navigate = useNavigate()
-    const config = getCurrentEcosystemPageOrComponentConfig('SSIDownloadPage') as SSIDownloadPageConfig
+    const flowRouter = useFlowRouter()
+    const config = flowRouter.getPageConfig() as SSIDownloadPageConfig
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/pages/SSIDownloadPage/index.tsx]: I guess we should add a generic to the function, so that it simply returns the type we need. const config: SSIDownloadPageConfig = flowRouter.getPageConfig() is nicer than having to do explicit casts everytime

File: packages/oid4vci-demo-frontend/src/router/flow-router.ts [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+type StepsByIdType = { [key: string]: VCIConfigRouteStep };
+
+interface StepState {
+    currentStep?: VCIConfigRouteStep
+    pageConfig?: PageConfig
+}
+
+export function useFlowAppRouter() {
+    const routes = getEcosystemRoutes()
+    const [currentRouteId, setCurrentRouteId] = useState<string>('')
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/router/flow-router.ts]: Why is it initialized with an empty string and not undefined or null?

File: packages/oid4vci-demo-frontend/src/router/flow-router.ts [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    const [stepsById] = useState<StepsByIdType>(buildStepsByIdMap(routes, getRouteId()))
+
+
+    function getRouteId(): string {
+        return currentRouteId != '' ? currentRouteId : 'default'
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/router/flow-router.ts]: See remark about empty string. Rather just have return currentRouteId ?? 'default'

File: packages/oid4vci-demo-frontend/src/router/flow-router.ts [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+export function useFlowRouter() {
+    const navigate = useNavigate()
+    const pageLocation = useLocation()
+    const routes = getEcosystemRoutes()
+    const [currentRouteId, setCurrentRouteId] = useState<string>('')
+    const [stepsById] = useState<StepsByIdType>(buildStepsByIdMap(routes, getRouteId()))
+    const stepState = useMemo<StepState>(() => initStepState(), [pageLocation.pathname])
+
+    function getRouteId(): string {
+        return currentRouteId != '' ? currentRouteId : 'default'
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/router/flow-router.ts]: Remark about empty strings now cascading to other functions

**Review by nklomp — CHANGES_REQUESTED** [review]
See remarks

File: packages/oid4vci-demo-frontend/src/ecosystem-config.ts (hunk 8 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/ecosystem-config.ts]: Not all steps have a page so also no page config (like create-credential-offer). 
Next commit will have hasCurrentEcosystemPageConfig()

File: packages/oid4vci-demo-frontend/src/pages/SSIDownloadPage/index.tsx (hunk 9 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/pages/SSIDownloadPage/index.tsx]: Fixed in next commit.

File: packages/oid4vci-demo-frontend/src/router/flow-router.ts (hunk 10 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/router/flow-router.ts]: I got an error on
```
const [currentRouteId, setCurrentRouteId] = useState<string|undefined>(undefined)
```
But I'll try just
```
const [currentRouteId, setCurrentRouteId] = useState<string>()
```
It compiles

File: packages/oid4vci-demo-frontend/src/router/flow-router.ts (hunk 11 above)
Comment (sanderPostma) [reply packages/oid4vci-demo-frontend/src/router/flow-router.ts]: okay
=>
```
const [currentRouteId, setCurrentRouteId] = useState<string>('default')
```

File: packages/oid4vci-demo-frontend/src/ecosystem-config.ts [hunk 13]
```diff
@@ -14,41 +14,60 @@ export function getCurrentEcosystemConfig(): VCIConfig {
...
-export function getEcosystemSequenceConfig(config?: VCIConfig): VCIConfigSequence {
+export function hasCurrentEcosystemPageConfig(stepId: string, config?: VCIConfig): boolean {
     if (!config) {
         config = getCurrentEcosystemConfig()
     }
-    return config.sequencer
+    return stepId in config.pages;
+}
+
+export function getCurrentEcosystemPageConfig(stepId: string, config?: VCIConfig): PageConfig | undefined{
```
Comment (nklomp) [inline packages/oid4vci-demo-frontend/src/ecosystem-config.ts]: No more undefined should be possible with current code

**Review by nklomp — COMMENTED** [review]
Smal remark about undefined being returned where it is not possible anymore. For the rest the PR can be merged after that change
